### PR TITLE
intel-oneapi-mkl: add missing compiler libraries for thread=openmp

### PR DESCRIPTION
**Pull request by albestro**
This is somehow a follow up of #37637, where MKL libraries related to threading were added, but some compiler-related libraries suggested by the link-line advisor were still missing.

This PR adds the missing threading-related compiler libraries to fix OpenMP support for both GNU and Intel compilers.

Changelog:
- ~main logic has been copy-pasted from old MKL~ rely on compiler to enable OpenMP as a follow up of #38719 (slightly different than what link-line advisor suggests).
- important detail: the `intel-oneapi-mkl` package logic has been slightly adapted in order to end up having the correct linking order, i.e. "{MKL} {MPI} {compiler} {system}".

### CODE REVIEW

File: lib/spack/spack/build_systems/oneapi.py [hunk 1]
```diff
@@ -195,6 +196,47 @@ def libs(self):
...
+                )
+            omp_libs = LibraryList(omp_lib_path.strip())
+
+        elif "%clang" in self.spec:
+            omp_libname = "libomp"
+            with self.compiler.compiler_environment():
+                omp_lib_path = Executable(self.compiler.cc)(
+                    "--print-file-name", f"{omp_libname}.{dso_suffix}", output=str
+                )
+            omp_libs = LibraryList(omp_lib_path.strip())
```
Comment (msimberg) [inline lib/spack/spack/build_systems/oneapi.py]: Let me take this opportunity to mention this PR, which changes the way OpenMP is added to intel packages: https://github.com/spack/spack/pull/38719. Instead of using library names/paths, I think this should be using the OpenMP flags, at least for clang. The original issue (https://github.com/spack/spack/issues/38509) has more details about what works and what doesn't. The issue was reported for `intel-mkl` but I'm guessing the same issue would be there for `intel-oneapi-mkl`.

Comment (albestro) [reply lib/spack/spack/build_systems/oneapi.py]: Thanks for pointing that out!

I agree with you, I quickly checked `clang` [user manual](https://releases.llvm.org/17.0.1/tools/clang/docs/UsersManual.html#openmp-features) for some of the recent releases  and they always state to use `-fopenmp`.

I would change `gcc` too, since that's what they report also in [their manual](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/libgomp/Enabling-OpenMP.html) (I checked a couple of latest versions starting from `gcc@10`).

**Comment by rscohn2** [comment]
> I agree with you, I quickly checked clang [user manual](https://releases.llvm.org/17.0.1/tools/clang/docs/UsersManual.html#openmp-features) for some of the recent releases and they always state to use -fopenmp.

I also prefer using the `-fopenmp` switch

File: lib/spack/spack/build_systems/oneapi.py (hunk 1 above)
Comment (albestro) [reply lib/spack/spack/build_systems/oneapi.py]: I had a look also at [intel compiler user manual](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-10/add-openmp-support.html) and it looks a bit trickier (see notes about macOS and LD_LIBRARY_PATH in [`-qopenmp`/`-fopenmp` flags](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-10/qopenmp-qopenmp.html#GUID-4AC3D696-F9A5-4064-A34A-A948300B80C7))

In addition to this, I'm trying to understand if from the package we can access the compiler property `openmp_flag`.

**Comment by albestro** [comment]
Probably went too far...let me know what you @rscohn2 @msimberg think about it. 😉

**Comment by rscohn2** [comment]
> Probably went too far...let me know what you @rscohn2 @msimberg think about it. 😉

It looks good to me. I will wait for others to comment before merging.

File: lib/spack/spack/build_systems/oneapi.py (hunk 1 above)
Comment (msimberg) [reply lib/spack/spack/build_systems/oneapi.py]: > I would change gcc too, since that's what they report also in [their manual](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/libgomp/Enabling-OpenMP.html) (I checked a couple of latest versions starting from gcc@10).

I realize now the one possible downside of using this approach is that when mixing compilers you may also end up mixing OpenMP runtimes. If an intel package is built with e.g. icc it would/should be using `libiomp`, but a dependent package built with e.g. GCC could end up with `libgomp` (since it doesn't explicitly specify which runtime to use, but just says `-fopenmp`). I suppose this might be one reason why things were set up the way they were.

However, the current state of this is still an improvement and I don't have a suggestion on how to better model this in spack.

**Review by msimberg — APPROVED** [review]

**Comment by albestro** [comment]
As @msimberg said, there might be cases where this is not going to work, e.g. mixing up compilers. It's not the only problem, for sure. For instance, as pointed out by @rasolca, this new approach also slightly change the behaviour, since we enable OpenMP not just at link-time but also at build-time.

I think there are for sure cases where this might not work, but they look like edge-cases (mixing compilers, and so OpenMP runtimes, does not sound like a good idea 😬 ), at least more than the one we are trying to fix with this PR.

TL;DR I would say is definitely far from being perfect, but I don't see any clear solution to the problem, and this PR might be a small improvement.

**Comment by msimberg** [comment]
@rscohn2 does this still look ok to you with the additional comments made? Should we go ahead with it in the current form?
